vtex-apps/glovo-integration
# Fix/product updates

#### What problem is this solving?
Each Glovo store's catalog is updated completely every hour.
To do this we have an endpoint that runs the `OrderForm's Simulation` for the products on the Glovo Menu in batches of 300 items at a time.
In the last version of the Integration we added the `minimumStock` property which is now being used in this simulation calls, instead of sending `quantity = 1`.
This created a problem where the simulation was returning a large portion of the items as `out of stock` or `unable to deliver`.

To solve this issue:
- Remove the hard-coded `max items per simulation` value and adding new GraphQL query (getMaxItemsForSimulation) and mutation (updateMaxItemsForSimulation) to allow the value to be changed if needed.
- Limit the `minimumStock` to be between 0 and 20.

We found another problem while debugging this issue. Because we are making multiple calls to the simulation, sometimes it returns an error.

To solve this issue: 
- Add a new utility function that wraps the request and retries the request after a short delay if it fails. The function takes three arguments: `request`, `retries` (defaults to 2) and `delay` (defaults to 500 milliseconds).

#### How to test it?
The integration is linked in this [workspace](https://glovo--ametllerorigenqa.myvtex.com)

You can fire a complete menu update by calling this endpoint:
```
POST: https://glovo--ametllerorigenqa.myvtex.com/_v/glovo/menu/update-all
```

After a few minutes, you can check the update for each store using this endpoint:
```
GET: https://app.io.vtex.com/vtex.glovo-integration/v3/ametllerorigenqa/glovo/_v/glovo/store-menu-updates/:affiliateId
```
(You can use `store-01` as affiliateId)

You should get a response like this one:
```
{
	"items": [
		{
			"id": "16",
			"available": true,
			"price": 0
		},
		...
	],
	"transactionId": null,
	"lastUpdated": "Mon Mar 20 2023 11:41:03 GMT+0000 (Coordinated Universal Time)",
	"successful": false,
	"minimumStock": 20
}
```
